### PR TITLE
doc/dev: add guide about submitting/reviewing PRs

### PR DESCRIPTION
**Pull request by benesch**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4244)

### CODE REVIEW

**Review by cirego — COMMENTED** [review]
This is really impressive and very well-written!

Meta-nit: It does appear that this PR violates the single RU principle, as there are multiple independent parts that are each worthy of discussion (why PRs, code style guide, testing guidelines, etc). However, I recognize that those conversations likely already happened and this is a merge of those conversations.

**Review by benesch — COMMENTED** [review]
Thanks very much for taking a look.

> Meta-nit: It does appear that this PR violates the single RU principle, as there are multiple independent parts that are each worthy of discussion (why PRs, code style guide, testing guidelines, etc). However, I recognize that those conversations likely already happened and this is a merge of those conversations.

I mean, it's a guide about what to look for in a PR and... those are all things about what to look for in a PR? I think of the "one semantic change" rule as meaning "don't add a feature and fix a bug in the same commit", since it becomes way harder to see what code is fixing the bug and what code is adding the feature, and way harder to revert just the bugfix or just the feature. What you're arguing against seems to be akin to arguing against a PR that adds a new feature that consists of three independent components? But that seems totally fine to me—possibly even preferable, because you get to see how all the components fit together—assuming the total size of the PR is still less than 500ish lines.

If anything I think this PR is pushing the size limit for prose, but I wasn't sure that splitting it into pieces would be easier ti review. If there's a way to restructure/chunk this guide up so it's a bit more digestible, definitely open to that!

File: doc/developer/guide-changes.md
Comment (cirego) [reply doc/developer/guide-changes.md:139]: I agree with your response entirely and find myself aligning closer to the wording in Google's guidelines.

What about something like the following?

> There is no "one-size fits all rule" for an acceptable number of modified lines in a PR and your reviewer may ask you to break your PR into smaller units, even if the original PR could be considered a single semantic change. In our experience, reviewers rarely request to break up PRs when fewer than 500 lines are changed and frequently request reviews broken into smaller PRs when the size exceeds 1000 lines.

File: doc/developer/guide-changes.md
Comment (cirego) [reply doc/developer/guide-changes.md:250]: You have changed my mind. Knowing what changes are trivial is a skill and definitely helps give people confidence in their work.

File: doc/developer/guide-changes.md
Comment (cirego) [reply doc/developer/guide-changes.md]: So, part of this comes my experience at Pure Storage, where `merge skew` was not the only cause for a broken `main`. There were a lot of long-running tests that could break and people would often "nominate someone else" without taking responsibility for fixing the test urgently. Hence why I believe it's important to escalate instead of nominate. Sending a message on Slack to say "Hey, I can't look now, but I'll look once available" works for me too.

If you believe that we can always run 100% of our tests on PRs before they merge, then I agree with what you've written. `merge skew` is sufficiently rare, and we are sufficiently small, that engineers will almost never by blocked by it.

File: doc/developer/guide-changes.md
Comment (ruchirK) [inline doc/developer/guide-changes.md:96]: this is a good place to throw in some sample commits that are good examples and maybe also link to the wiki page for imperative tense?

File: doc/developer/guide-changes.md
Comment (ruchirK) [inline doc/developer/guide-changes.md:393]: that seems pretty low latency. I'm not against this in theory, but I don't think we are meeting this latency bound very often. If this is meant to be aspirational - totally fine and maybe worth rewording? If this is meant as the actual policy - I think it might need to be socialized more?

File: doc/developer/guide-changes.md
Comment (ruchirK) [inline doc/developer/guide-changes.md:8]: Would be nice to call out the places we intentionally deviate strongly from the google guide, if there are any

File: doc/developer/guide-changes.md
Comment (ruchirK) [inline doc/developer/guide-changes.md:345]: [potentially out of scope]: everything in this section talks about how the code reviewer helps the reviewee who is new to Rust / new to this part of the codebase. But imo doing code reviews is also very beneficial to the code reviewer as well. 

In my experience, reviewing the work of engineers more and less senior than me helps me understand how to better think about code, reviewing changes in a part of the codebase I am unfamiliar with helps me ramp up on that part, and doing code reviews helps me if in the future there are bugs / customer issues with that part of the code.

I suspect that none of this is controversial, but I think it might be nice to include in this doc somewhere (doesn't have to be here) to help motivate those readers who might be less inclined to do code reviews (especially if we hire less experienced people in the future)?

File: doc/developer/guide-changes.md
Comment (ruchirK) [inline doc/developer/guide-changes.md:200]: I think it would be nice to also ask that engineers self-review their code (however they want) before submitting it for review by others? I've found that this often catches small things that would otherwise be needless cognitive overhead for the reviewer

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:139]: SGTM!

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:200]: Good idea! Added a section called "polish requirements" that covers this.

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:96]: Done.

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:345]: Good idea! Added to the "Choosing reviewers" section up above.

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:438]: Yeah, proposing this guideline is actually the catalyst for this entire document. I think this is possibly the single most important thing we can do to increase the velocity of the entire engineering org.

Comment (benesch) [reply doc/developer/guide-changes.md:438]: Can definitely socialize this more before this merges.

File: doc/developer/guide-changes.md
Comment (ruchirK) [reply doc/developer/guide-changes.md:211]: kind of in a similar vein -- i personally love when people call out the parts of their changes they feel most unsure about in the pr description - i know opinion is a little split on that. If you think its worth doing, might be nice to also add that to the guide?

File: doc/developer/guide-changes.md
Comment (mjibson) [reply doc/developer/guide-changes.md]: A note about bors use at Cockroach (currently in the many 10s of engineers): it was indeed suspected that bors caused more problems than it solved, so it was disabled. During the next day 5 merge skews occurred in master and it was thus re-enabled because it was evident that it was useful. I suggest having some data-backed trigger, like X merge skews per day/week and we look into using bors.

File: doc/developer/guide-changes.md
Comment (mjibson) [reply doc/developer/guide-changes.md:295]: Yay I'm so happy for this paragraph!

**Review by mjibson — APPROVED** [review]

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:8]: Just one minor one. Added a section on this point.

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:211]: Sure, done!

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md]: Oh, that's helpful context, thanks Matt! Just spitballing, but merge skew once a week seems like about the right time to use Bors? I put that in for now, but definitely subject to debate.

Comment (mjibson) [reply doc/developer/guide-changes.md]: Yes? I think it makes sense to use bors if (time that bors saves preventing merge skews and its associated problems) > (time it takes to implement and perform maintenance on the bors service). 1/week might be there? Really hard to say.

File: doc/developer/guide-changes.md
Comment (quodlibetor) [inline doc/developer/guide-changes.md:556]: I would add to this that minimizing complexity is part of adding features, because it makes it easier to add future features -- they're not orthogonal.

File: doc/developer/guide-changes.md
Comment (quodlibetor) [inline doc/developer/guide-changes.md:704]: Also with perfect testing *not* introducing a bug should only ever require *adding* tests.

File: doc/developer/guide-changes.md
Comment (quodlibetor) [inline doc/developer/guide-changes.md:736]: nit: if this is referring to gerrit's "change list is a list of changes", this philosophy comes from the fact that gerrit makes reviewing individual commits trivial and correct while github makes it impossible. If you want to review a branch merge in gerrit, the *only* way to do it (from my experience, possibly we had a customized install) is by reviewing each commit ("change") individually.

So philosophically I'd say that gerrit is just the semantic-commits philsophy, enabled at a technical level.

**Review by quodlibetor — APPROVED** [review]

File: doc/developer/guide-changes.md
Comment (JLDLaughlin) [inline doc/developer/guide-changes.md:294]: Could we add a subsection on how to actually merge? (More specifically, how to rebase interactively versus the questionable Github "squash and merge" button?)

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:736]: Hmm, maybe I have no idea what I'm talking about then. I'm just going to strike the last sentence in this paragraph in the interest of correctness.

**Comment by benesch** [comment]
Added the two items we talked about in the meeting today:

  * PRs stuck in review should be downgraded back to drafts.
  * No review expectations on PRs with failing tests.

I think that's everything! Going to merge EOD tomorrow, so get your last round of feedback in now! (Future feedback is very much welcomed, though I will be retiring from wordsmithing duties; you'll have to draw up the PR with your proposed change yourself.)

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:741]: I'm a bit confused by this phrasing. If you have perfect testing, then by definition you don't need to add any tests?

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:593]: Done.

File: doc/developer/guide-changes.md
Comment (benesch) [reply doc/developer/guide-changes.md:305]: There's a bit on using the "squash" strategy in the "Commit strategy" section, but not sure whether that really reflects your lived experience. Would you be up for making the adjustments you see fit once this merges?

Comment (JLDLaughlin) [reply doc/developer/guide-changes.md:305]: Yes!

**Review by cirego — APPROVED** [review]

**Pull request merged**
